### ISSUE TITLE
Saving My Work after clicking Save or Finish button

### ISSUE THREAD

**Issue by alexrcabral** [benetech/MathShare#1284]
Expected behavior: Work which is not intentionally added should not be added as a step. Any changes made should be retained within the My Work and step description box when saving or finishing, and any applicable cleanup should remain in the step previous (and retained as an "change" to the My Work section).

[Screenshot of a problem 3x plus 4 equals 7. The second step has crossouts which cleanup to 3x equals 3. 3x equals 3 is similarly shown in the My Work box.] 

E.g. If a student has 3x=3 in the my work section which matches the prior cleanup stage, this would be there after saving. If they changed it to x=1 then saved, the x=1 would be in the My Work section upon their return. 

The exception to this is if the student is sending a view only version of the set, such as to turn in a homework assignment. In this scenario, work shown in the math editor that is different from the final step added should be displayed as a final step. This is consistent with students’ mental model of the “My Work” section being the current step, or in this case the final step.

E.g. If a student had changed it to x=1 before saving or finishing, then when they turn in the work, this would be shown as Step 3: x=1. 

This does have the small discrepancy of allowing for there to not be a description on the final step of a given turned in problem. This is acceptable for now, since this fits more closely to the existing paradigms, while preventing loss of work.

Future considerations not covered in this issue: 
-What all should be handled with “Finishing” a step (e.g. converting to view only, which could allow for requirement of step description for final step)

**Comment by johnhbenetech** [comment]
@alexrcabral an edge case/bug that came up on a recent call with @rupeshparab 

Scenario:
* A student has added 5 steps already
* A student goes back to edit step 2
* A student does NOT complete the edit (by clicking 'update step')
* The student hits Finish or tries to save/navigate away with the uncommited edit still in the my work box

What should happen? Two obvious solutions I can imagine:
1. The finish/save/navigate away causes the edit to overwrite the existing step 2
2. The finish/save/navigate away throws out the changes and leaves step 2 unchanged

**Comment by alexrcabral** [comment]
I don't see how this differs from the editing of any other step. To continue the same paradigm within this scenario:

> Any changes made should be retained within the My Work and step description box when saving or finishing

Any changes made should be retained within the My Work, which would remain beneath Step 2 (with the other steps located beneath it). The idea is if a student is saving their work, they should be able to continue where they left off.

**Comment by johnhbenetech** [comment]
Ok thanks @alexrcabral can you confirm my understanding?

@rupeshparab So it sounds like we may need to make some tweaks to the implementation we discussed on the last call. The 'in progress' step probably needs to include some concept of step number/ordering. So to reiterate, 'in progress' steps can be:

* after the last step is added, any changes a user makes in the my work box that differs from that last added step should be saved as an 'in progress' step. This value should be loaded into the 'my work' box when a user returns. If a user uses the share my work read only link, it should render the in progress step as a numbered step.

* if the user has an edit of a previous step in progress when they save/finish/leave a page, if the uncommitted work differs from the existing step this should also be saved as an 'in progress' step appropriately linked to the step it is modifying. When a user returns to that problem, it should load as if the edit for that step was still in progress with the last state populated in the my work box.

@alexrcabral what happens if the user clicks 'share my work' with an edit in progress? Should it render the existing step, or the edit in progress step in the readonly version?

**Comment by alexrcabral** [comment]
Thanks @johnhbenetech, that's exactly right. 

The share my work with an edit in progress should render the edit in progress step just as it does for the final step. 

>The exception to this is if the student is sending a view only version of the set, such as to turn in a homework assignment. In this scenario, work shown in the math editor that is different from the final step added should be displayed as a final step.

In this scenario, given that the student is sending a view only version of the set, such as to turn in a homework assignment, work shown in the math editor that is different from the in progress step (e.g. Step 2) should be displayed as the updated step.

**Comment by johnhbenetech** [comment]
ok cool.

@rupeshparab please let us know if there are additional questions.

**Comment by rupeshparab** [comment]
Will this progress step also be available for teachers when they are creating problem sets
Also, in case it is, and teacher is sharing the problem set with steps, will the progress step be shown to students in the shared version?

**Comment by johnhbenetech** [comment]
@alexrcabral can you comment on the question above?

**Comment by alexrcabral** [comment]
@rupeshparab To clarify, user type (Teacher vs Student) is irrelevant to the anticipated workflow. 

What is of import is if the user has 
1) Started solving a problem
2) Has made recent changes while solving the problem *without clicking to add or update the step*
3) Has attempted to save said problem

To the question of sharing problem with steps, yes, the set should be shared as it was left, and retain the same logic as when students are sharing the view only set. In this scenario, given that a user is sharing their work with steps, such as to assign a partially filled out problem set, work shown in the math editor that is different from the last-edited step added should be displayed in the final shared version.

**Comment by bognasoldev89** [comment]
@rupeshparab @alexrcabral I noticed a few things during testing this issue. I followed scenarios and the basic remark is when the user provides something new only to the left part of "my work" and leaves the commentary part empty, nothing is saved.

1.
Example: I'm logged in as a teacher, I added a few steps, I started developing my equation and put nothing into a comment. Next, I clicked "Next problem", had to save changes again by modal, and went back to the problem developed earlier. No changes were applied.
![ms 1284 1](https://user-images.githubusercontent.com/51324797/88923903-bf8cc580-d272-11ea-8274-0380a21454f7.gif)

Below the example with commentary section filled-in, the "my work" section is saved:
![ms 1284 2](https://user-images.githubusercontent.com/51324797/88924067-037fca80-d273-11ea-8300-095dcb7e7607.gif)

2.
As I understood, using the clean-up should work a little differently - the following input should not be saved. But with comment provided, the work is saved as in the first scenario:
![ms 1284 3](https://user-images.githubusercontent.com/51324797/88924298-59ed0900-d273-11ea-94de-d0d3e6e13394.gif)

3.
Saving "my work" in the middle step works fine, it always goes back to the right step:
![ms 1284 4](https://user-images.githubusercontent.com/51324797/88924481-a33d5880-d273-11ea-8c8e-491465ad271e.gif)

I have 2 remarks:
- Deleting commentary section, leaving problem page, and going back results in no changes, "My work" section is placed under the last step and no changes are applied to the "almost-changed" step.
- After successfully coming back to opened "almost-changed" middle step and clicking "Cancel edit", those changes are still present in the "My work" section below the last step. I suppose values from the last step should be available there.
![ms 1284 5](https://user-images.githubusercontent.com/51324797/88924978-3e363280-d274-11ea-8577-ebd66049ad09.gif)

4.
And, to sum up, 5 - both of these scenarios are fulfilled if commentary section is filled-in.
![ms 1284 6](https://user-images.githubusercontent.com/51324797/88926442-4abb8a80-d276-11ea-98a6-922bcee2f2e8.gif)

Additionally, I would like to get your attention on this discard/save modal. In case when the user is opening a problem with steps provided already, adding anything to the "My work" section without adding a new step and changes page - it seems unnecessary. I'll leave it here for another discussion, maybe in #968 

To sum up:
1. Commentary section seems to determine if "My work" is saved or not
2. After clean-up steps behavior is different as it should be (@alexrcabral please confirm)
3. Wrong behavior after canceling editing the middle step - wrong input in "My work" section

**Comment by johnhbenetech** [comment]
Thanks @bognasoldev89 !

@rupeshparab please look at 1 & 3

@alexrcabral please comment on 2

**Comment by rupeshparab** [comment]
I have pushed fixes for issues 1 and 3

**Comment by johnhbenetech** [comment]
QA, please test 1&3

@alexrcabral please comment on 2

**Comment by bognasoldev89** [comment]
So it seems like 1 & 3 are resolved.
Commentary section doesn't determine saving "My work" and remarks I had about editing middle steps don't occur anymore.

One thing I noticed - while sharing answers/problem sets with clean-up steps in "My work", the step before the clean-up step is not displayed.

![ms 1284 cleanup 1](https://user-images.githubusercontent.com/51324797/89508931-e7c67800-d7ce-11ea-97f7-973e20dcc4fa.png)

![ms 1284 cleanup 2](https://user-images.githubusercontent.com/51324797/89508935-e8f7a500-d7ce-11ea-8c08-162d7bdc214a.png)

**Comment by alexrcabral** [comment]
@bognasoldev89 I'm a bit unclear on your question here, but on testing it's performing as expected. 

The changes made are correctly retained within the My Work and step description box when saving or finishing, and any applicable cleanup correctly remains in the step previous & is retained as a change in the My Work section.

**Comment by alexrcabral** [comment]
@rupeshparab This follow-up point does need to be addressed: 
> One thing I noticed - while sharing answers/problem sets with clean-up steps in "My work", the step before the clean-up step is not displayed.

When this final step is added (shown in the math editor), it should be added as a Step 3 (along with applicable cleanup) rather than replacing the existing Step. 

Including @bognasoldev89's screenshots for reference: 

![Screenshot showing an editable version with steps 1 and 2, and editing currently taken place in the editor.](https://user-images.githubusercontent.com/51324797/89508931-e7c67800-d7ce-11ea-97f7-973e20dcc4fa.png)

![Screenshot showing a view-only version with steps 1 and 2. However, the step 2 displayed was the editor content from the previous screenshot, and the old step 2 appears to be missing.](https://user-images.githubusercontent.com/51324797/89508935-e8f7a500-d7ce-11ea-8c08-162d7bdc214a.png)

**Comment by rupeshparab** [comment]
@bognasoldev89 
Can you check if you can replicate this issue?
I tried replicating it, the final step with Calc seems to be added to the review along with cleanup

This is the sample which I tried:
Edit link: https://mathshare-qa.diagramcenter.org/#/app/problem/edit/CWTOJEHAOICLG
![image](https://user-images.githubusercontent.com/6897505/90175622-4ee2be80-ddc5-11ea-8e0e-c09b1c0b8f28.png)

View Link: https://mathshare-qa.diagramcenter.org/#/app/problem/view/45P63ZTQL7PVG
![image](https://user-images.githubusercontent.com/6897505/90175660-5c984400-ddc5-11ea-889e-43feb6ade0e3.png)

**Comment by johnhbenetech** [comment]
@alexrcabral can you actually look at this. @bognasoldev89 is on holiday tomorrow

**Comment by alexrcabral** [comment]
@rupeshparab @johnhbenetech 

Sure thing -- I was able to recreate this using the same problem and steps as @bognasoldev89 using Chrome on macOS (10.15.5) 

https://www.loom.com/share/cc8b65842b344ad7bbcba56a8c51e8ce

**Comment by bognasoldev89** [comment]
@alexrcabral 

> @bognasoldev89 I'm a bit unclear on your question here, but on testing it's performing as expected.
The changes made are correctly retained within the My Work and step description box when saving or finishing, and any applicable cleanup correctly remains in the step previous & is retained as a change in the My Work section.

I believe this "any further work within the My Work section should be retained" from the OP mislead me. But thanks for the clarification!
> Scenario 2: Given that I'm a user who has recently used a button that results in a cleanup step (e.g. calculate, cross-out), when I press "Save" or "Finish", then the clean-up step should remain, and any further work within the My Work section should be retained.

**Comment by rupeshparab** [comment]
@bognasoldev89 
I have pushed a fix for the issue where the new in progress step wasn't shown in review, please verify it
Note: you will need to generate a new review link as there was a problem with its generation previously

**Comment by bognasoldev89** [comment]
@rupeshparab it seems fine now.

- Original
![Screenshot from 2020-08-19 13-05-28](https://user-images.githubusercontent.com/51324797/90627869-e492a980-e21c-11ea-9a66-d217b37bde52.png)

- Shared set
![Screenshot from 2020-08-19 13-05-57](https://user-images.githubusercontent.com/51324797/90627874-e6f50380-e21c-11ea-9151-26ef14dc204b.png)

- Shared answers
![Screenshot from 2020-08-19 13-06-22](https://user-images.githubusercontent.com/51324797/90627886-e9eff400-e21c-11ea-87de-873c5342bdaf.png)

@alexrcabral @johnhbenetech one more thing occurred to me, regarding crossing out
So step by step:
1. I'm resolving the equation
2. I'm using cross-out/cross-out and replace
3. I'm not adding a step
4. I'm leaving problem's page
5. I'm going back to the problem's page
6. I want to revert crossing out

Shouldn't this be possible to achieve? If this only one step provided, clear all should work, but in the case of more steps I wouldn't want to lose all my work.

![Peek 2020-08-19 13-03](https://user-images.githubusercontent.com/51324797/90627668-8c5ba780-e21c-11ea-9f70-cb68c0253a22.gif)

**Comment by johnhbenetech** [comment]
This last case should be taken into account as part of #1356
